fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/synonym/SynonymMapRegistryImpl.java

```diff
@@ -0,0 +1,122 @@
+import java.nio.charset.StandardCharsets;
+import java.text.ParseException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.ServiceLoader;
+
+/**
+ * Registry for {@link SynonymMap}s.
+ */
+public class SynonymMapRegistryImpl implements SynonymMapRegistry {
+    @Nonnull
+    private static final Logger LOGGER = LoggerFactory.getLogger(SynonymMapRegistryImpl.class);
+
+    private static final SynonymMapRegistryImpl INSTANCE = new SynonymMapRegistryImpl();
+
+    @Nonnull
+    private Map<String, SynonymMap> registry;
+
+    private SynonymMapRegistryImpl() {
+        registry = initRegistry();
+    }
+
+    @Nonnull
+    public static SynonymMapRegistry instance() {
+        return INSTANCE;
+    }
+
+    @Nonnull
+    @Override
+    public SynonymMap getSynonymMap(@Nonnull final String name) {
+        final SynonymMap map = registry.get(name);
+        if (map == null) {
+            throw new MetaDataException("unrecognized synonym map", LogMessageKeys.SYNONYM_NAME, name);
+        }
+        return map;
+    }
+
+    @Override
+    public synchronized void register(final String name, @Nonnull final InputStream synonymInputStream) {
+        SynonymMap oldMap = registry.putIfAbsent(name, buildSynonymMap(new InputStreamReader(synonymInputStream, StandardCharsets.UTF_8)));
+        if (oldMap != null) {
+            throw new RecordCoreArgumentException("attempted to register duplicate synonym analyzer", LogMessageKeys.SYNONYM_NAME, name);
+        }
+        if (LOGGER.isInfoEnabled()) {
+            LOGGER.info(KeyValueLogMessage.of("registered tokenizer", LogMessageKeys.SYNONYM_NAME, name));
+        }
+    }
+
+    @Override
+    public void reset() {
+        registry = initRegistry();
+    }
+
+    @Nonnull
+    private static Map<String, SynonymMap> initRegistry() {
+        final Map<String, SynonymMap> registry = new HashMap<>();
+        for (SynonymMapConfig config : ServiceLoader.load(SynonymMapConfig.class)) {
+            registry.put(config.getName(), buildSynonymMap(new InputStreamReader(config.getSynonymInputStream(), StandardCharsets.UTF_8)));
+
+        }
+        return registry;
+    }
+
+    private static SynonymMap buildSynonymMap(final InputStreamReader synonymStream) {
+        try {
+            SynonymMap.Parser parser = new SolrSynonymParser(true, true, new Analyzer() {
+                @Override
+                protected TokenStreamComponents createComponents(String fieldName) {
+                    final StandardTokenizer src = new StandardTokenizer();
+                    TokenStream tok = new LowerCaseFilter(src);
+                    tok = new FlattenGraphFilter(tok);
+                    return new TokenStreamComponents(src, tok);
+                }
+            });
+            parser.parse(synonymStream);
+            return parser.build();
+        } catch (IOException | ParseException ex) {
+            throw new RecordCoreException("Failed to parse wordnet for synonym analyzer", ex);
```

Comment:
Is this exception message sufficiently general, i.e. should it be about "wordnet"?

Reply:
`buildSynonymMap` should probably take the name, and add it as log info, otherwise, users with more than one synonym map, won't know which one failed to parse.

Reply:
:+1: